File: services/shared/schemas/permission.ts
import { z } from "zod";

// Permission matrix: role × resource × action × allow/deny.
//
// Resources are strings (not enum) so new resource types can land without
// a schema migration. Action is a fixed enum — adding an action requires
// thinking through default policy for every existing role.
//
// `clientId` scopes a permission to a single client when the role grants
// limited access (e.g. `client_guest`). `null` means workspace-wide.

export const PermissionActionSchema = z.enum([
  "read",
  "create",
  "update",
  "delete",
  "approve",
  "deny",
  "publish",
  "invite",
  "revoke",
  "export",
  "admin",
]);
export type PermissionAction = z.infer<typeof PermissionActionSchema>;

export const PermissionResourceSchema = z.string().min(1).max(60).regex(/^[a-z0-9_]+$/);
export type PermissionResource = z.infer<typeof PermissionResourceSchema>;

// The standard resources the platform ships with. Custom resource strings
// are accepted by the schema; we enumerate the standard set here so callers
// have an authoritative reference.
export const STANDARD_RESOURCES = [
  "company",
  "user",
  "membership",
  "role",
  "permission",
  "agent",
  "lesson",
  "draft",
  "approval",
  "audit_log",
  "meter_event",
  "billing",
  "integration",
  "brand_kit",
  "campaign",
  "channel",
] as const;
export type StandardResource = (typeof STANDARD_RESOURCES)[number];

export const PermissionSchema = z.object({
  id: z.string().uuid(),
  companyId: z.string().uuid(),
  roleId: z.string().uuid(),
  resource: PermissionResourceSchema,
  action: PermissionActionSchema,
  allow: z.boolean().default(true),
  clientId: z.string().uuid().nullable().optional(),
  createdAt: z.string().datetime(),
});
export type Permission = z.infer<typeof PermissionSchema>;

export const PermissionCheckRequestSchema = z.object({
  userId: z.string().uuid(),
  companyId: z.string().uuid(),
  clientId: z.string().uuid().nullable().optional(),
  resource: PermissionResourceSchema,
  action: PermissionActionSchema,
});
export type PermissionCheckRequest = z.infer<typeof PermissionCheckRequestSchema>;

export const PermissionCheckResponseSchema = z.object({
  allowed: z.boolean(),
  reason: z.enum([
    "matched_allow",
    "matched_deny",
    "no_matching_rule",
    "no_membership",
    "membership_revoked",
  ]),
  matchedPermissionId: z.string().uuid().nullable().optional(),
});
export type PermissionCheckResponse = z.infer<typeof PermissionCheckResponseSchema>;

File: services/shared/schemas/permission.py
"""Permission schema — Python mirror of permission.ts."""

from __future__ import annotations

import re
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

PermissionAction = Literal[
    "read",
    "create",
    "update",
    "delete",
    "approve",
    "deny",
    "publish",
    "invite",
    "revoke",
    "export",
    "admin",
]

STANDARD_RESOURCES: tuple[str, ...] = (
    "company",
    "user",
    "membership",
    "role",
    "permission",
    "agent",
    "lesson",
    "draft",
    "approval",
    "audit_log",
    "meter_event",
    "billing",
    "integration",
    "brand_kit",
    "campaign",
    "channel",
)

_RESOURCE_RE = re.compile(r"^[a-z0-9_]+$")


class Permission(BaseModel):
    id: str
    company_id: str
    role_id: str
    resource: str = Field(..., min_length=1, max_length=60)
    action: PermissionAction
    allow: bool = True
    client_id: str | None = None
    created_at: datetime

    @field_validator("resource")
    @classmethod
    def _resource_format(cls, v: str) -> str:
        if not _RESOURCE_RE.match(v):
            raise ValueError("resource must match ^[a-z0-9_]+$")
        return v


class PermissionCheckRequest(BaseModel):
    user_id: str
    company_id: str
    client_id: str | None = None
    resource: str = Field(..., min_length=1, max_length=60)
    action: PermissionAction

    @field_validator("resource")
    @classmethod
    def _resource_format(cls, v: str) -> str:
        if not _RESOURCE_RE.match(v):
            raise ValueError("resource must match ^[a-z0-9_]+$")
        return v


CheckReason = Literal[
    "matched_allow",
    "matched_deny",
    "no_matching_rule",
    "no_membership",
    "membership_revoked",
]


class PermissionCheckResponse(BaseModel):
    allowed: bool
    reason: CheckReason
    matched_permission_id: str | None = None
